### PR TITLE
Status messages rebase

### PR DESCRIPTION
**Pull request by edewit**
- [x] Have you created an [issue](https://github.com/redhat-kontinuity/catapult/issues) and used it in the commit message?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/redhat-kontinuity/catapult/pulls) for the same issue?
- [x] Have you built the project locally prior to submission with `mvn clean install`?

-----

rebased version of status messages branch that adds a websocket endpoint for status messages

### CODE REVIEW

File: tests/pom.xml [hunk 1]
```diff
@@ -97,6 +97,16 @@
       <version>1.0.2</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.glassfish.tyrus</groupId>
+      <artifactId>tyrus-client</artifactId>
+      <version>1.1</version>
```
Comment (gastaldi) [inline tests/pom.xml]: Do we need this or just for tests? If the latter, shouldn't it be in test scope?

File: core/core-api/src/main/java/org/kontinuity/catapult/core/api/StatusMessageEvent.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package org.kontinuity.catapult.core.api;
```
Comment (gastaldi) [inline core/core-api/src/main/java/org/kontinuity/catapult/core/api/StatusMessageEvent.java]: This package name was changed when we renamed to missioncontrol

File: core/core-impl/src/main/java/io/openshift/appdev/missioncontrol/core/impl/MissionControlImpl.java [hunk 3]
```diff
@@ -92,16 +100,24 @@ public Boom launch(CreateProjectile projectile) throws IllegalArgumentException
         File path = projectile.getProjectLocation().toFile();
         String repositoryDescription = projectile.getGitHubRepositoryDescription();
         GitHubRepository gitHubRepository = gitHubService.createRepository(projectName, repositoryDescription);
+        event.fire(new StatusMessageEvent(projectile.getId(), StatusMessage.GITHUB_CREATE, singletonMap("location", gitHubRepository.getHomepage())));
```
Comment (gastaldi) [inline core/core-impl/src/main/java/io/openshift/appdev/missioncontrol/core/impl/MissionControlImpl.java]: What happens if an exception is thrown when performing an operation? Is the websocket endpoint also notified?

File: tests/pom.xml (hunk 1 above)
Comment (edewit) [inline tests/pom.xml]: this whole module is called `test` will it still need scope test?

Comment (gastaldi) [inline tests/pom.xml]: The other dependencies are also in test scope, so I'd say yes :)

File: core/core-impl/src/main/java/io/openshift/appdev/missioncontrol/core/impl/MissionControlImpl.java (hunk 3 above)
Comment (edewit) [inline core/core-impl/src/main/java/io/openshift/appdev/missioncontrol/core/impl/MissionControlImpl.java]: Your project will never be done.. don't know if / how we could improve that. I think error handling as a whole needs to be improved first

File: core/core-impl/src/main/java/io/openshift/appdev/missioncontrol/core/impl/MissionControlImpl.java [hunk 4]
```diff
@@ -92,16 +99,24 @@ public Boom launch(CreateProjectile projectile) throws IllegalArgumentException
...
         OpenShiftProject createdProject = openShiftService.createProject(projectName);
+        event.fire(new StatusMessageEvent(projectile.getId(), StatusMessage.OPENSHIFT_CREATE, singletonMap("location", createdProject.getConsoleOverviewUrl())));
         openShiftService.configureProject(createdProject, gitHubRepository.getGitCloneUri());
 
         GitHubWebhook webhook = getGitHubWebhook(gitHubService, openShiftService, gitHubRepository, createdProject);
+        fireEvent(projectile, StatusMessage.GITHUB_WEBHOOK);
         return new BoomImpl(gitHubRepository, createdProject, webhook);
     }
 
+    private void fireEvent(Projectile projectile, StatusMessage status) {
```
Comment (gastaldi) [inline core/core-impl/src/main/java/io/openshift/appdev/missioncontrol/core/impl/MissionControlImpl.java]: Remove this method and use `event.fire` directly

File: web/src/main/java/io/openshift/appdev/missioncontrol/web/api/websocket/MissionControlStatusEndpoint.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * A websocket based resource that informs clients about the status of the operations
+ *
+ * https://abhirockzz.wordpress.com/2015/02/10/integrating-cdi-and-websockets/
+ */
+@ServerEndpoint(value = "/status/{uuid}")
+public class MissionControlStatusEndpoint {
+    private static final Logger log = Logger.getLogger(MissionControlStatusEndpoint.class.getName());
+
+    private static Map<UUID, Session> peers = Collections.synchronizedMap(new WeakHashMap<>());
```
Comment (gastaldi) [inline web/src/main/java/io/openshift/appdev/missioncontrol/web/api/websocket/MissionControlStatusEndpoint.java]: Why `WeakHashMap`?

Comment (edewit) [inline web/src/main/java/io/openshift/appdev/missioncontrol/web/api/websocket/MissionControlStatusEndpoint.java]: we don't want to keep the sessions from being garbage collected in the event that we didn't receive the onclose.

File: core/core-impl/src/main/java/io/openshift/appdev/missioncontrol/core/impl/MissionControlImpl.java (hunk 3 above)
Comment (edewit) [inline core/core-impl/src/main/java/io/openshift/appdev/missioncontrol/core/impl/MissionControlImpl.java]: added error handling now as well

File: web/src/main/java/io/openshift/appdev/missioncontrol/web/api/websocket/MissionControlStatusEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * A websocket based resource that informs clients about the status of the operations
+ *
+ * https://abhirockzz.wordpress.com/2015/02/10/integrating-cdi-and-websockets/
+ */
+@ServerEndpoint(value = "/status/{uuid}")
+public class MissionControlStatusEndpoint {
+    private static final Logger log = Logger.getLogger(MissionControlStatusEndpoint.class.getName());
+
+    private static Map<UUID, Session> peers = Collections.synchronizedMap(new HashMap<>());
```
Comment (gastaldi) [inline web/src/main/java/io/openshift/appdev/missioncontrol/web/api/websocket/MissionControlStatusEndpoint.java]: Collections.synchronizedMap can be replaced with ConcurrentHashMap

Comment (gastaldi) [inline web/src/main/java/io/openshift/appdev/missioncontrol/web/api/websocket/MissionControlStatusEndpoint.java]: See http://stackoverflow.com/questions/510632/whats-the-difference-between-concurrenthashmap-and-collections-synchronizedmap

Comment (gastaldi) [inline web/src/main/java/io/openshift/appdev/missioncontrol/web/api/websocket/MissionControlStatusEndpoint.java]: And made it final

**Comment by quintesse** [comment]
Squash-merged. Closing

**Pull request closed without merging**